### PR TITLE
[front] - feat(members): search members endpoint

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description

This PR implements pagination search for the member list in the `CreateOrEditVaultModal` component.

It introduces a new `/search` endpoint for members, which allows for efficient server-side pagination and filtering of workspace members. The changes improve performance and user experience when creating or editing vaults, especially for workspaces with a large number of members.

**References:**
- https://github.com/dust-tt/tasks/issues/1284

## Risk

Minor as it is only used in vaults

## Deploy Plan

- Deploy `front-edge`
- Deploy `front`

### CODE REVIEW

**Comment by JulesBelveze** [comment]
Note: will all use this endpoint within the `MemberList` component. But I would do it in a separate PR as I would also like to refactor it to use `DataTable`

File: front/pages/api/w/[wId]/members/search.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+              "Invalid query parameters. Both 'searchTerm' and 'orderBy' must be strings.",
+          },
+        });
+      }
+
+      const paginationRes = getPaginationParams(req, {
+        defaultLimit: DEFAULT_PAGE_LIMIT,
+        defaultOrderColumn: orderBy || "name",
+        defaultOrderDirection: "desc",
+        supportedOrderColumn: ["createdAt", "email", "name"],
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/members/search.ts:52]:
```suggestion
        supportedOrderColumn: ["name"],
```

File: front/lib/resources/user_resource.ts [hunk 2]
```diff
@@ -143,6 +145,42 @@ export class UserResource extends BaseResource<User> {
     return user ? new UserResource(User, user.get()) : null;
   }
 
+  static async searchUsers(
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:148]:
```suggestion
  static async listUsersWithEmailPredicat(
```

File: front/lib/resources/user_resource.ts [hunk 3]
```diff
@@ -143,6 +145,42 @@ export class UserResource extends BaseResource<User> {
...
+    workspaceId: number,
+    options: {
+      searchTerm?: string;
```
Comment (flvndvd) [inline front/lib/resources/user_resource.ts:151]: Let's reflect that it search on email

File: front/lib/api/workspace.ts [hunk 4]
```diff
@@ -177,6 +178,43 @@ export async function getMembers(
...
+
+  const { users, total } = await UserResource.searchUsers(
+    owner.id,
+    {
+      searchTerm: options.searchTerm,
+    },
+    paginationParams
+  );
+
+  const usersJson = users.map((u) => {
```
Comment (flvndvd) [inline front/lib/api/workspace.ts:201]: user.toJSON() 😛

File: front/components/vaults/CreateOrEditVaultModal.tsx [hunk 5]
```diff
@@ -84,8 +96,43 @@ export function CreateOrEditVaultModal({
     }
   }, [vaultMembers]);
 
-  const router = useRouter();
-  const sendNotification = useContext(SendNotificationsContext);
+  const debouncedFetch = useCallback(
```
Comment (flvndvd) [inline front/components/vaults/CreateOrEditVaultModal.tsx:99]: Let's move this to a hook here: https://github.com/dust-tt/dust/blob/main/front/lib/swr/user.ts

File: front/components/vaults/CreateOrEditVaultModal.tsx [hunk 6]
```diff
@@ -61,8 +63,18 @@ export function CreateOrEditVaultModal({
     vault?.name ?? null
   );
   const [selectedMembers, setSelectedMembers] = useState<string[]>([]);
+  // const [members, setMembers] = useState<UserType[]>([]);
```
Comment (flvndvd) [inline front/components/vaults/CreateOrEditVaultModal.tsx:66]: ??

File: front/components/vaults/CreateOrEditVaultModal.tsx [hunk 7]
```diff
@@ -61,8 +63,18 @@ export function CreateOrEditVaultModal({
...
   const [searchTerm, setSearchTerm] = useState<string>("");
   const [isSaving, setIsSaving] = useState(false);
+  // const [isLoading, setIsLoading] = useState(false);
```
Comment (flvndvd) [inline front/components/vaults/CreateOrEditVaultModal.tsx:69]: ??

File: front/components/vaults/CreateOrEditVaultModal.tsx [hunk 8]
```diff
@@ -97,14 +113,25 @@ export function CreateOrEditVaultModal({
...
+        enableSorting: false,
       },
       {
         id: "action",
         cell: (info: Info) => {
           const isSelected = selectedMembers.includes(info.row.original.userId);
           if (isSelected) {
             return (
-              <div className="full-width flex justify-end">
+              <div className="full-width ml-4 flex justify-end">
```
Comment (flvndvd) [inline front/components/vaults/CreateOrEditVaultModal.tsx:134]: 😱

File: front/lib/api/workspace.ts [hunk 9]
```diff
@@ -177,6 +178,32 @@ export async function getMembers(
...
+    {
+      email: options.email,
+    },
+    paginationParams
+  );
+
+  const usersJson = users.map((u) => {
+    return u.toJSON();
+  });
+  return { members: usersJson, total };
```
Comment (flvndvd) [inline front/lib/api/workspace.ts:204]:
```suggestion
  return { members: users.map((u) => u.toJSON()), total };
```

File: front/lib/swr/user.ts [hunk 10]
```diff
@@ -30,3 +33,44 @@ export function useUserMetadata(key: string) {
...
+  const { data, error, mutate } = useSWRWithDefaults(
+    `/api/w/${workspaceId}/members/search?${searchParams.toString()}`,
+    searchMembersFetcher,
+    {
+      revalidateOnFocus: false,
+      revalidateOnReconnect: false,
+    }
+  );
+  return {
+    members: data?.members ?? [],
```
Comment (flvndvd) [inline front/lib/swr/user.ts:70]:
```suggestion
    members: useMemo(() => (data ? data.members : []), [data]),
```

File: front/lib/swr/user.ts [hunk 11]
```diff
@@ -30,3 +33,44 @@ export function useUserMetadata(key: string) {
...
+    membersCount: data?.total ?? 0,
```
Comment (flvndvd) [inline front/lib/swr/user.ts:71]: We already have `totalNodesCount` can we stick to the same naming so `totalMembersCount`

**Review by flvndvd — APPROVED** [review]

File: front/components/vaults/CreateOrEditVaultModal.tsx (hunk 6 above)
Comment (JulesBelveze) [reply front/components/vaults/CreateOrEditVaultModal.tsx:66]: forgot to push last commit sorry..

**Pull request merged**